connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java

```diff
@@ -145,154 +166,79 @@ public void poll(long timeout) {
-                if (memberTasks == null) {
-                    memberTasks = new ArrayList<>();
-                    taskAssignments.put(taskAssignedTo, memberTasks);
-                }
-                memberTasks.add(taskId);
+            log.debug("Deserialized new assignment: {}", newAssignment);
+            if (assignmentSnapshot != null) {
+                assignmentSnapshot.connectors().removeAll(newAssignment.revokedConnectors());
+                assignmentSnapshot.tasks().removeAll(newAssignment.revokedTasks());
+                log.debug("Snapshot after delete assignment: {}", assignmentSnapshot);
+                newAssignment.connectors().addAll(assignmentSnapshot.connectors());
+                newAssignment.tasks().addAll(assignmentSnapshot.tasks());
             }
+            log.debug("Augmented new assignment: {}", newAssignment);
         }
-
-        this.leaderState = new LeaderState(memberConfigs, connectorAssignments, taskAssignments);
-
-        return fillAssignmentsAndSerialize(memberConfigs.keySet(), ConnectProtocol.Assignment.NO_ERROR,
-                leaderId, memberConfigs.get(leaderId).url(), maxOffset, connectorAssignments, taskAssignments);
+        assignmentSnapshot = newAssignment;
+        listener.onAssigned(assignmentSnapshot, generation);
     }
 
-    private Map<String, ByteBuffer> fillAssignmentsAndSerialize(Collection<String> members,
-                                                                short error,
-                                                                String leaderId,
-                                                                String leaderUrl,
-                                                                long maxOffset,
-                                                                Map<String, List<String>> connectorAssignments,
-                                                                Map<String, List<ConnectorTaskId>> taskAssignments) {
-
-        Map<String, ByteBuffer> groupAssignment = new HashMap<>();
-        for (String member : members) {
-            List<String> connectors = connectorAssignments.get(member);
-            if (connectors == null)
-                connectors = Collections.emptyList();
-            List<ConnectorTaskId> tasks = taskAssignments.get(member);
-            if (tasks == null)
-                tasks = Collections.emptyList();
-            ConnectProtocol.Assignment assignment = new ConnectProtocol.Assignment(error, leaderId, leaderUrl, maxOffset, connectors, tasks);
-            log.debug("Assignment: {} -> {}", member, assignment);
-            groupAssignment.put(member, ConnectProtocol.serializeAssignment(assignment));
-        }
-        log.debug("Finished assignment");
-        return groupAssignment;
+    @Override
+    protected Map<String, ByteBuffer> performAssignment(String leaderId, String protocol, List<JoinGroupResponseMember> allMemberMetadata) {
+        return currentConnectProtocol == CONNECT_PROTOCOL_V0
```

Comment:
this is another location where I wonder if we just maintain one assignor and update *it* when we see protocol version change rather than maintaining `currentConnectProtocol` and switching on it could refactor some of this logic to be more general

Reply:
See comment above about protocol classes being effectively static.